src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.State.cs

```diff
@@ -0,0 +1,204 @@
+                Document = document;
+                RootNamespace = rootNamespce;
+                RelativeTargetNamespace = relativeTargetNamespace;
+                TargetNamespace = targetNamespace;
+                DeclaredNamespace = declaredNamespace;
+                RelativeDeclaredNamespace = relativeDeclaredNamespace;
+            }
+
+            private static bool IsLinkedDocument(Document document)
+            {
+                var solution = document.Project.Solution;
+
+                // TODO: figure out how to properly determine if a document is linked using project system.
+
+                // If we found a linked document which is part of a project with differenct project file,
+                // then it's an actual linked file (i.e. not a MTFM project).
+                if (document.GetLinkedDocumentIds()
+                    .Any(id => !PathUtilities.PathsEqual(solution.GetDocument(id).Project.FilePath, document.Project.FilePath)))
+                {
+                    return true;
+                }
+                
+                // Now determine if the file path match its logical path in workspace,
+                // only trigger the refactoring when they match.
+                var projectRoot = PathUtilities.GetDirectoryName(document.Project.FilePath);
+                var folderPath = Path.Combine(document.Folders.ToArray());
+
+                var absoluteDircetoryPath = PathUtilities.GetDirectoryName(document.FilePath);
+                var logicalDirectoryPath = PathUtilities.CombineAbsoluteAndRelativePaths(projectRoot, folderPath);
+
+                return !PathUtilities.PathsEqual(absoluteDircetoryPath, logicalDirectoryPath);
+            }
+
+            public static async Task<State> CreateAsync(TService service, Document document, TextSpan textSpan, CancellationToken cancellationToken)
+            {
+                if (document.Project.FilePath == null)
+                {
+                    return null;
+                }
+
+                if (!textSpan.IsEmpty)
+                {
+                    return null;
+                }
+
+                var workspace = document.Project.Solution.Workspace;
+                if (workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+                {
+                    return null;
+                }
+
+                if (document.IsGeneratedCode(cancellationToken))
+                {
+                    return null;
+                }
+
+                // Bail if we can't get the root namespace for the project, the refactoring depends on it.
+                var rootNamespace = document.Project.DefaultNamespace;
+                if (rootNamespace == null)
+                {
+                    return null;
+                }
+
+                // Ignore linked documents
+                // TODO: this only detect documents linked multiple times, need to figure out how to detect a document is linked, e.g. using <Link>, or shared project, etc.
```

Comment:
1. wrap lines.
2. explain why it would matter to undertand "how" the document is linked.